### PR TITLE
Feature/saving ideas in concept

### PR DESCRIPTION
**Pull request by LorenzoJokhan**
# Description
Added migration to add publishedDate field to the ideas table
Added migration to update all old plans with a publishedField value equal to their creation date
Changed onlyVisible to only return ideas that are in concept when created by yourself or if you are an admin.

Please include
Is a dependency of https://github.com/openstad/openstad-frontend/pull/359

### CODE REVIEW

File: src/models/Idea.js [hunk 1]
```diff
@@ -514,24 +521,38 @@ module.exports = function (db, sequelize, DataTypes) {
...
+            where: {
+              [Op.or]: [
+                {
+                  [Op.or]: [
+                    {userId},
+                    {viewableByRole: 'all'},
+                    {viewableByRole: null},
+                    {viewableByRole: roles[userRole] || ''}
+                  ],
+                  publishedDate: {[Op.ne]: null}
```
Comment (nlsvgtr) [inline src/models/Idea.js:541]: I could be reading this wrong, but I think this means that an Idea that has viewableByRole set to admin is now viewable by other users when the published date is null.

Comment (nlsvgtr) [reply src/models/Idea.js:541]: But maybe it works correctly because Sequelize defaults this to an AND statement?

**Review by nlsvgtr — CHANGES_REQUESTED** [review]
I think it's ok like this. I added one comment where I am not sure, please double check.

One request: if it's all the same to you could you rename publishedDate to publishDate? That would give us the option to add extra functionality later: if you would use dates in the future you could write something now that would be automatically be published later.

File: src/models/Idea.js (hunk 1 above)
Comment (LorenzoJokhan) [reply src/models/Idea.js:541]: Sequealize creates an AND statement between each array of [Op.or]. When you are not logged in than you will only receive the published ideas that you are allowed to view, this is the query that is generated for when the user is logged in:

SELECT
    `idea`.`id`,
    `idea`.`siteId`,
    `idea`.`userId`,
    `idea`.`startDate`,
    `idea`.`sort`,
    `idea`.`typeId`,
    `idea`.`status`,
    `idea`.`viewableByRole`,
    `idea`.`title`,
    `idea`.`summary`,
    `idea`.`description`,
    `idea`.`budget`,
    `idea`.`extraData`,
    `idea`.`location`,
    `idea`.`modBreak`,
    `idea`.`modBreakUserId`,
    `idea`.`modBreakDate`,
    `idea`.`publishedDate`,
    `idea`.`createdAt`,
    `idea`.`updatedAt`,
    `idea`.`deletedAt`,
    (
        SELECT
            COUNT(*)
        FROM
            votes v
        WHERE
            v.deletedAt IS NULL
            AND (
                v.checked IS NULL
                OR v.checked = 1
            )
            AND v.ideaId = idea.id
            AND v.opinion = "yes"
    ) AS `yes`,
    (
        SELECT
            COUNT(*)
        FROM
            votes v
        WHERE
            v.deletedAt IS NULL
            AND (
                v.checked IS NULL
                OR v.checked = 1
            )
            AND v.ideaId = idea.id
            AND v.opinion = "no"
    ) AS `no`,
    (
        SELECT
            COUNT(*)
        FROM
            arguments a
        WHERE
            a.deletedAt IS NULL
            AND a.ideaId = idea.id
    ) AS `argCount`,
    `tags`.`id` AS `tags.id`,
    `tags`.`name` AS `tags.name`
FROM
    `ideas` AS `idea`
    LEFT OUTER JOIN (
        `ideaTags` AS `tags->ideaTags`
        INNER JOIN `tags` AS `tags` ON `tags`.`id` = `tags->ideaTags`.`tagId`
    ) ON `idea`.`id` = `tags->ideaTags`.`ideaId`
    AND (`tags`.`deletedAt` IS NULL)
WHERE
    (
        `idea`.`deletedAt` IS NULL
        AND (
            (
                `idea`.`viewableByRole` = 'all'
                OR `idea`.`viewableByRole` IS NULL
                OR `idea`.`viewableByRole` = ''
            )
            AND NOT (`idea`.`publishedDate` IS NULL)
            AND `idea`.`siteId` = '3'
        )
    )
ORDER BY
    `idea`.`createdAt` DESC;

**Comment by LorenzoJokhan** [comment]
Changed all ocurences of publishedDate to publishDate, will test tomorrow if everything still works and will give a sign if everything is ok

File: src/models/Idea.js [hunk 2]
```diff
@@ -307,7 +298,23 @@ module.exports = function (db, sequelize, DataTypes) {
...
     },
-
+    publishDate: {
+      type: DataTypes.DATE,
+      allowNull: true
+    },
+    publishDateHumanized: {
+      type: DataTypes.VIRTUAL,
+      get: function () {
+        var date = this.getDataValue('publishDate');
```
Comment (rudivanhierden) [inline src/models/Idea.js:308]:
```suggestion
        const date = this.getDataValue('publishDate');
```

File: src/routes/api/idea.js [hunk 3]
```diff
@@ -173,6 +173,12 @@ router.route('/')
     let userId = req.user.id;
     if (req.user.role == 'admin' && req.body.userId) userId = req.body.userId;
 
+    if(!req.body.publishAsConcept) {
+      req.body['publishDate'] = new Date();
+    } else {
+      req.body['publishDate'] = null;
+    }
```
Comment (rudivanhierden) [inline src/routes/api/idea.js:180]: This code is now duplicated in this file, can you extract it to a middleware?

**Review by rudivanhierden — CHANGES_REQUESTED** [review]

File: migrations/036-prefill-existing-plans-published-date.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+var db = require('../src/db').sequelize;
+
+module.exports = {
+	up: function() {
+		try {
+			return db.query(`
+				UPDATE ideas 
+				SET publishedDate = createdAt
+				WHERE createdAt < '2022-09-30'
```
[the remark this answers is not in the document]
Comment (LorenzoJokhan) [reply migrations/036-prefill-existing-plans-published-date.js:9]: That is true, i will remove the where

**Review by rudivanhierden — APPROVED** [review]
Thanks for the changes, LGTM! 🚀

File: src/models/Idea.js (hunk 1 above)
Comment (LorenzoJokhan) [reply src/models/Idea.js]: Everything works as expected

**Pull request merged**